src/components/legend/draw.js

```diff
@@ -184,27 +184,51 @@ module.exports = function draw(gd) {
     if(anchorUtils.isRightAnchor(opts)) {
         lx -= opts.width;
     }
-    if(anchorUtils.isCenterAnchor(opts)) {
+    else if(anchorUtils.isCenterAnchor(opts)) {
         lx -= opts.width / 2;
     }
 
     if(anchorUtils.isBottomAnchor(opts)) {
         ly -= opts.height;
     }
-    if(anchorUtils.isMiddleAnchor(opts)) {
+    else if(anchorUtils.isMiddleAnchor(opts)) {
         ly -= opts.height / 2;
     }
 
+    //lx = Math.round(lx);
+    //ly = Math.round(ly);
+
+    // Make sure the legend top and bottom are visible
+    // (legends with a scroll bar are not allowed to stretch beyond the extended
+    // margins)
+    var lyMin = 0;
+    var lyMax = fullLayout.margin.t + fullLayout.height + fullLayout.margin.b;
+    var legendHeight = opts.height;
+    var legendHeightMax = gs.h;
+
+
+    if(legendHeight > legendHeightMax) {
+        ly = gs.t;
+        legendHeight = legendHeightMax;
+    }
+    else {
+        if(ly > lyMax) ly = lyMax - legendHeight;
+
+        if(ly < lyMin) ly = lyMin;
+
+        legendHeight = Math.min(lyMax - ly, opts.height);
+    }
+
     // Deal with scrolling
-    var plotHeight = fullLayout.height - fullLayout.margin.b,
-        scrollheight = Math.min(plotHeight - ly, opts.height),
-        scrollPosition = scrollBox.attr('data-scroll') ? scrollBox.attr('data-scroll') : 0;
+    var scrollPosition = scrollBox.attr('data-scroll') ?
```

Comment:
For conditional assignments like this, we can make it a bit more concise/readable:

``` js
var scrollPosition = scrollBox.attr('data-scroll') || 0;
```

and use ternary operators in situations where the conditional value isn''t used except as the condition.